ProjFS.Mac/PrjFSKext/PrjFSKext/AssociativeArray.hpp

```diff
@@ -0,0 +1,203 @@
+#pragma once
+
+#include "Locks.hpp"
+#include "Memory.hpp"
+#include "KextCppUtilities.hpp"
+
+struct ResizeResult
+{
+    bool   memoryWasUsed;
+    uint32_t releasedMemorySize;
+    void*  releasedMemory;
+};
+
+template <typename KEY, typename VALUE, bool EQUAL_FN(const KEY&, const VALUE&)>
+class AssociativeArray
+{
+    AssociativeArray(const AssociativeArray&) = delete;
+    AssociativeArray& operator=(const AssociativeArray&) = delete;
+
+    uint32_t capacity;
+    uint32_t numberUsed;
+    VALUE* data;
+    
+    uint32_t NextGrowCapacity() const
+    {
+        return (this->capacity == 0) ? 8u : (this->capacity * 2u);
+    }
+    
+public:
+    AssociativeArray() :
+        capacity(0),
+        numberUsed(0),
+        data(nullptr)
+    {}
+    
+    ~AssociativeArray()
+    {
+        this->ClearAndFreeMemory();
+    }
+    
+    void Clear()
+    {
+        for (uint32_t i = 0; i < this->numberUsed; ++i)
+        {
+            this->data[i].~VALUE();
+        }
+        this->numberUsed = 0;
+    }
+    
+    void ClearAndFreeMemory()
+    {
+        this->Clear();
+        
+        if (nullptr != this->data)
+        {
+            Memory_Free(this->data, MemorySizeForCapacity(this->capacity));
+            this->data = nullptr;
+            this->capacity = 0;
+        }
+    }
+    
+    uint32_t MemorySizeToGrow() const
+    {
+        return this->NextGrowCapacity() * sizeof(this->data[0]);
+    }
+    
+    static uint32_t MemorySizeForCapacity(uint32_t capacity)
+    {
+        return capacity * sizeof(VALUE);
+    }
+    
+    ResizeResult GrowUsingMemory(void* memory, uint32_t memorySize)
+    {
+        uint32_t expectedCapacity = this->NextGrowCapacity();
+        uint32_t expectedMemorySize = this->MemorySizeToGrow();
+        if (memorySize != expectedMemorySize)
+        {
+            return ResizeResult{ false };
+        }
+        
+        return this->ReserveCapacity(expectedCapacity, memory, memorySize);
+    }
+
+    ResizeResult ReserveCapacity(uint32_t newCapacity, void* memory, uint32_t memorySize)
+    {
+        assert(MemorySizeForCapacity(newCapacity) == memorySize);
+        assert(newCapacity >= this->numberUsed);
+        
+        uint8_t* memoryPosition = static_cast<uint8_t*>(memory);
+        for (size_t i = 0; i < this->numberUsed; ++i, memoryPosition += sizeof(VALUE))
+        {
+            VALUE* moved = new(memoryPosition) VALUE(KextCpp::move(this->data[i]));
+            (void)moved;
+            this->data[i].~VALUE();
+        }
+        
+        ResizeResult result = { true, MemorySizeForCapacity(this->capacity), this->data };
+        this->data = static_cast<VALUE*>(memory);
+        this->capacity = memorySize / sizeof(VALUE);
```

Comment:
should we be able to just do
this->capacity = newCapacity;

Reply:
Yes, good point. Saves an integer divide. :-)